eclipse-che/che-server
# feat: native auth on Kubernetes

Signed-off-by: Michal Vala <mvala@redhat.com>



### What does this PR do?
 - removed `che.infra.kubernetes.enable_unsupported_k8s` + any checks to it
 - Implemented `OIDCTokenInitializationFilter` that checks token from Authorization header, decodes it and creates Che user from token claims.
 - Moved multiple `Keycloak*` classes into new OIDC module and keep generic OIDC stuff there. Where needed, Keycloak implementation extends these OIDC ones.
 - Moved namespace configuration into `NamespaceConfigurator`s. This makes `KubernetesNamespaceFactory` and `OpenShiftProjectFactory` simpler.
 - Move calling of `NamespaceConfigurator` from `NamespaceProvisioner` into `KubernetesNamespaceFactory`/`OpenShiftProjectFactory`. This makes factories slightly more complex, but still better than doing all configuration/provisioning directly by Factory. 


### Screenshot/screencast of this PR



### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/20633


### How to test this PR?

che-server image: `quay.io/mvala/che-server:gh20633-nativeAuthKube`
che-operator image: `quay.io/mvala/che-operator:gh20635-nativeAuthKube`

1. start minikube
2. create CheCluster `patch.yaml`, replace `{{MINIKUBE_IP}}` with your minikube IP:
  ```
spec:
    auth:
      nativeUserMode: true
      identityProviderURL: http://dex.dex:5556
      gatewayAuthenticationSidecarImage: quay.io/oauth2-proxy/oauth2-proxy:v7.2.0
      gatewayAuthorizationSidecarImage: quay.io/brancz/kube-rbac-proxy:v0.11.0
    server:
      cheHost: 'che-eclipse-che.{{MINIKUBE_IP}}.nip.io'
      customCheProperties:
        CHE_KEYCLOAK_AUTH__INTERNAL__SERVER__URL: 'http://dex.dex:5556'
```
3. deploy with devworkspaces engine and images (you'll have to accept the risk that chectl prompts) ^^ 
  `chectl server:deploy --platform=minikube --installer=operator --cheimage=quay.io/mvala/che-server:gh20633-nativeAuthKube --che-operator-image=quay.io/mvala/che-operator:gh20635-nativeAuthKube --workspace-engine=dev-workspace --che-operator-cr-patch-yaml=patch.yaml`
4. wait until che is fully deployed (there should be no keycloak), then you can open che url as usual and import the CA cert
5. You get into Dex login page. There are couple of predefined users. `che@eclipse.org:admin` or `userN@che:password` where N is 1-5. After you login, you should get to dashboard as you know.



### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
